nft_ingester/src/consistency_bg_job.rs

```diff
@@ -0,0 +1,743 @@
+async fn run_peers_checking_loop(
+    storage: Arc<Storage>,
+    peers_provider: Arc<dyn AuraPeersProvides + Send + Sync>,
+    force_reingestable_slot_processor: Arc<dyn BlockConsumer + Send + Sync>,
+    metrics: Arc<Peer2PeerConsistencyMetricsConfig>,
+) {
+    let mut last_processed_epoch = 0u32;
+
+    loop {
+        let next_processing_slot = calc_exchange_slot_for_epoch(last_processed_epoch + 1);
+        let current_slot = last_tracked_slot();
+        if current_slot < next_processing_slot {
+            let duration = slots_to_time(next_processing_slot - current_slot);
+            tokio::time::sleep(duration).await;
+            continue;
+        }
+
+        let epoch_to_process = epoch_of_slot(next_processing_slot);
+
+        let bbgm_task = {
+            let storage = storage.clone();
+            let peers_provider = peers_provider.clone();
+            let force_reingestable_slot_processor = force_reingestable_slot_processor.clone();
+            let metrics = metrics.clone();
+            tokio::spawn(async move {
+                exchange_bbgms_with_peers(
+                    epoch_to_process,
+                    storage,
+                    peers_provider,
+                    force_reingestable_slot_processor,
+                    metrics,
+                )
+                .await
+            })
+        };
+        let acc_task = {
+            let storage = storage.clone();
+            let peers_provider = peers_provider.clone();
+            let metrics = metrics.clone();
+            tokio::spawn(async move {
+                exchange_account_with_peers(storage, peers_provider, metrics.as_ref()).await
+            })
+        };
+        let _ = bbgm_task.await;
+        let _ = acc_task.await;
+
+        last_processed_epoch = epoch_to_process;
+    }
+}
+
+/// Exchanges bubblegum checksum with peers to identify missing bubblegum change,
+/// and requests these missing changes from peers.
+///
+/// ## Args:
+/// * epoch - epoch we want the checksums to be exchanged
+/// * storage - local database
+/// * peers_provider - source of trusted peers
+async fn exchange_bbgms_with_peers(
+    epoch: u32,
+    storage: Arc<Storage>,
+    peers_provider: Arc<dyn AuraPeersProvides + Send + Sync>,
+    force_reingestable_slot_processor: Arc<dyn BlockConsumer + Send + Sync>,
+    metrics: Arc<Peer2PeerConsistencyMetricsConfig>,
+) {
+    tracing::info!("Starting bubblegum changes peer-to-peer exchange for epoch={epoch}");
+    while get_calculating_bbgm_epoch()
+        .map(|e| e == epoch)
+        .unwrap_or(false)
+    {
+        tokio::time::sleep(Duration::from_secs(10)).await;
+    }
+
+    let grand_epoch = grand_epoch_of_epoch(epoch);
+
+    let mut missing_bbgm_changes: HashMap<BbgmChangeRecord, HashSet<usize>> = HashMap::new();
+    let trusted_peers = peers_provider.list_trusted_peers().await;
+
+    for (peer_ind, trusted_peer) in trusted_peers.iter().enumerate() {
```

Comment:
Maybe we need to try to make this process more async? Here many i/o with other peers, so maybe it makes sense to spawn separate tasks for communicating with each peer?

Reply:
This can be a good improvement in the future, but for the "phase 1", I'd prefer to run it sequentially, and collect CPU and disk metrics to understand whether we have some potential bandwidth for parallelization.
Cause we can easily fall into a scenario when we are stressing the disk with bunch of concurrent peer-2-peer consistency checks, and as the result the indexing part lacks resources for fulfilling it's main duty.